package gcm;

message Host {
    optional string name = 1;
    optional string inner_ip = 2;
    optional string outer_ip = 3;
    optional string deploy_ip = 4;
}

message HostCfg {
    repeated Host hosts = 1;
}

message ArtifactFile {
    optional string src = 1;
    optional string dst = 2;
    optional bool is_dir = 3;
    optional bool need_backup = 4;
    repeated string tags = 5;
}

message AgentArtifact {
    repeated ArtifactFile files = 1;
    optional string name = 2;
    optional int32 function_id = 3;
    optional string start_cmd = 4;
    optional string stop_cmd = 5;
    optional string clean_cmd = 6;
    optional string check_cmd = 7;
}

message TemplateArtifactItem {
    optional string template_name = 1;
    repeated ArtifactFile files = 2;
}

message TemplateArtifact {
    repeated TemplateArtifactItem template_artifacts = 1;
}

message NormalArtifactItem {
    optional string template_name = 1;
    optional string name = 2;
    optional int32 function_id = 3;
    optional int32 start_priority = 4;
    repeated ArtifactFile files = 5;
    optional uint32 copies = 6 [default = 1];
}

message NormalArtifact {
    repeated NormalArtifactItem artifacts = 1;
}

message ArtifactGroupItem {
    optional string name = 1;
    repeated string artifact_names = 2;
}

message ArtifactGroup {
    repeated ArtifactGroupItem artifact_groups = 1;
}

message KeyValue {
    optional string key = 1;
    optional string value = 2;
}

message InstanceGroup {
    optional string artifact_group_name = 1;
    optional string host_name = 2;
    optional int32 instance_id = 3;
}

message Zone {
    optional string name = 1;
    optional int32 id = 2;
    repeated InstanceGroup instance_groups = 4;
}

message World {
    optional string name = 1;
    optional int32 id = 2;
    optional string user = 3;
     optional string passwd = 4;
    repeated KeyValue variables = 5;
    repeated Zone zones = 6;
}

message DeployInfo {
    optional string tmp_root_path = 2;
    optional string dst_root_path = 3;
    optional int32 listen_port = 4;
}

message Deploy {
    optional string tmp_root_path = 2;
    optional string dst_root_path = 3;
    repeated KeyValue variables = 4;
    repeated World worlds = 5;
    optional int32 listen_port = 6;
}

